server/src/main/java/com/cloud/resource/ResourceManagerImpl.java

```diff
@@ -1353,46 +1353,57 @@ protected boolean setHostIntoMaintenance(HostVO host) throws NoTransitionExcepti
+        s_logger.debug("Host " + host.getUuid() + " entering in PrepareForMaintenance state as any previous corrections have been fixed");
+        resourceStateTransitTo(host, ResourceState.Event.ErrorsCorrected, _nodeId);
+        return true;
+    }
+
     /**
      * Return true if host goes into Maintenance mode, only when:
      * - No Running, Migrating or Failed migrations (host_id = last_host_id) for the host
      */
-    protected boolean attemptMaintain(HostVO host) throws NoTransitionException {
+    private boolean attemptMaintain(HostVO host) throws NoTransitionException {
         final long hostId = host.getId();
 
         if (CollectionUtils.isEmpty(_vmDao.findByHostInStates(hostId, State.Migrating, State.Running, State.Starting, State.Stopping, State.Error, State.Unknown))) {
             return setHostIntoMaintenance(host);
         }
 
         final List<VMInstanceVO> allVmsOnHost = _vmDao.listByHostId(hostId);
-        final List<VMInstanceVO> migratingVms = _vmDao.listVmsMigratingFromHost(hostId);
-        final List<VMInstanceVO> failedMigrations = _vmDao.listNonMigratingVmsByHostEqualsLastHost(hostId);
-        boolean hasPendingWorkForVMs = false;
+        boolean hasPendingMigrationWorks = false;
         for (VMInstanceVO vmInstanceVO : allVmsOnHost) {
-            if (_haMgr.hasPendingHaWork(vmInstanceVO.getId())) {
-                hasPendingWorkForVMs = true;
+            if (_haMgr.hasPendingMigrationsWork(vmInstanceVO.getId())) {
+                hasPendingMigrationWorks = true;
                 break;
             }
         }
 
-        if (!hasPendingWorkForVMs) {
-            if ((CollectionUtils.isNotEmpty(_vmDao.findByHostInStates(hostId, State.Running)) && CollectionUtils.isEmpty(migratingVms)) ||
-                    (CollectionUtils.isEmpty(_vmDao.findByHostInStates(hostId, State.Running)) &&
-                        CollectionUtils.isNotEmpty(_vmDao.findByHostInStates(hostId, State.Unknown, State.Error, State.Shutdowned)))) {
-                return setHostIntoErrorInMaintenance(host, failedMigrations);
-            }
+        final List<VMInstanceVO> failedMigrations = new ArrayList<>(_vmDao.listNonMigratingVmsByHostEqualsLastHost(hostId));
+        final List<VMInstanceVO> errorVms = new ArrayList<>(_vmDao.findByHostInStates(hostId, State.Unknown, State.Error, State.Shutdowned));
+        final boolean hasMigratingVms = CollectionUtils.isNotEmpty(_vmDao.listVmsMigratingFromHost(hostId));
+        final boolean hasRunningVms = CollectionUtils.isNotEmpty(_vmDao.findByHostInStates(hostId, State.Running));
+        final boolean hasFailedMigrations = CollectionUtils.isNotEmpty(failedMigrations);
+        final boolean hasVmsInFailureStates = CollectionUtils.isNotEmpty(errorVms);
+        errorVms.addAll(failedMigrations);
+
+        if (!hasPendingMigrationWorks && (hasRunningVms || (!hasRunningVms && !hasMigratingVms && hasVmsInFailureStates))) {
```

Comment:
Maybe this could be a separate method with a javadoc to explain it? To improve readability

Reply:
Moving to another method.